lib/os/onoff.c

```diff
@@ -650,3 +661,185 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
 
 	return rv;
 }
+
+#if CONFIG_ONOFF_DELAYED
+
+static void timeout_handler(struct _timeout *t);
+
+static struct onoff_delayed_client *get_delayed_cli(sys_snode_t *node)
+{
+	struct onoff_client *cli =
+			CONTAINER_OF(node, struct onoff_client, node);
+
+	 return CONTAINER_OF(cli, struct onoff_delayed_client, cli);
+}
+
+/**
+ * @brief Convert input timeout to an absolute timeout.
+ *
+ * Input timeout can be relative or absolute. It is reduced by the offset.
+ *
+ * @param timeout output timeout.
+ * @param t input timeout as provided by the user.
+ * @param offset offset given in system ticks.
+ *
+ * @retval true if request is instant.
+ * @retval false if request is in the future.
+ */
+static bool get_abs_time(k_timeout_t *timeout, k_timeout_t t, uint32_t offset)
+{
+	if (K_TIMEOUT_EQ(t, K_NO_WAIT)) {
+		return true;
+	}
+
+	if (Z_TICK_ABS(t.ticks) < 0) {
+		if (t.ticks < offset) {
+			return true;
+		}
+
+		timeout->ticks =
+			Z_TICK_ABS((z_tick_get() + t.ticks - offset));
+	} else {
+		timeout->ticks =
+			Z_TICK_ABS(Z_TICK_ABS(t.ticks) - offset);
+	}
+
+	return false;
+}
+
+sys_snode_t *sortlist_head_peek(sys_slist_t *list)
+{
+	return sys_slist_peek_head(list);
+}
+
+sys_snode_t *sortlist_get(sys_slist_t *list)
+{
+	return sys_slist_get(list);
+}
+
+bool sortlist_remove(sys_slist_t *list, sys_snode_t *node)
+{
+	return sys_slist_find_and_remove(list, node);
+}
+
+void sortlist_add(sys_slist_t *list, sys_snode_t *node)
+{
+	struct onoff_delayed_client *newcli = get_delayed_cli(node);
+	struct onoff_delayed_client *cli;
+	struct onoff_delayed_client *prev = NULL;
+	sys_snode_t *n;
+
+	SYS_SLIST_FOR_EACH_NODE(list, n) {
+		cli = get_delayed_cli(n);
+		/* ticks are negative (abs) so inverted comparison */
+		if (newcli->timeout.ticks > cli->timeout.ticks) {
```

Comment:
Comparison of absolute timeouts needs to be wrapped in API that abstracts from the hack of using a sign bit to identify an absolute timeout.